src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+﻿using System.Diagnostics.Contracts;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ProducerConsumerStream : Stream
+    {
+        private TaskCompletionSource<WriteBufferWrapper> _buffer;
+        private TaskCompletionSource<int> _dataAvail;
+        private WriteBufferWrapper _currentBuffer;
+        private bool _disposed;
+
+        public ProducerConsumerStream()
+        {
+            _buffer = new TaskCompletionSource<WriteBufferWrapper>();
+            _dataAvail = new TaskCompletionSource<int>();
+            _currentBuffer = WriteBufferWrapper.EmptyContainer;
+        }
+
+        public override void Flush()
+        {
+            _dataAvail.TrySetResult(0);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                return 0;
+            }
+
+            _buffer.SetResult(new WriteBufferWrapper(buffer, offset, count));
+            int totalBytesRead = 0;
+            do
+            {
+                int bytesRead = _dataAvail.Task.Result;
+                _dataAvail = new TaskCompletionSource<int>();
+                totalBytesRead += bytesRead;
+                count -= bytesRead;
+                if (bytesRead == 0)
+                {
+                    break;
+                }
+            } while (count > 0);
+            return totalBytesRead;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
```

Comment:
I think we either need to opt all-in for things like Contract.EndContractBlock() or opt all-out.  Once you opt in, then we probably need Contract.Requires and Contract.Assert on internal methods as well. I personally like use of Contract, but only if consistently embraced.  Enabling static analysis without an all-in adoption yields only noise.  Your call to leave this in or not.